BlockBasic Property changes

ToMe25 opened the pull request:
## Changes proposed in this pull request:
* make most BlockBasics Opaque for Compatibility(with for example Storage Drawers Framed Drawers)
* make some more BlockBasics full blocks
* change the BlockStateBasic default for full block to true
* Fix BlockBasic Resistance
* Reduce BlockBasic Resistance to match Iron Block
* Reduce Bronze Block, Steel Block, Thermal Evaporation Stuff and Security Desk Resistance to be Iron Block x1.5
* keep Osmium Block Resistance to be Iron Block x2
* Increase Refined Obsidian Hardness to match Obsidian
* Increase Osmium Block Hardness to be Iron Block x 1.5

thiakil commented on src/main/java/mekanism/common/block/states/BlockStateBasic.java:
```diff
@@ -210,6 +218,14 @@ public boolean hasRotations() {
         }
+        
+        public boolean isOpaqueCube() {
+            if(meta == 0 && !loaded) {//This is needed to prevent Security Desks and Structural Glass from getting Opaque somhow.
```
I'd much rather you solved this properly

ToMe25 replied:
I tried to get it without that but than all blocks are Opaque idk why.

thiakil replied:
Please define "all blocks are opaque" - how do you know this? what observable effect does it have?

ToMe25 replied:
Structural Glass and Security Desk are getting rendered wrong.
Like black and the glass is doing too much shadows.

thiakil replied:
alright, so have you set a breakpoint here (inside your `if` statement), and see what is calling it?

ToMe25 replied:
yeah i have tried that and its the Block init.

ToMe25 replied:
ok i got that now i am trying to fix the explosion resistance of all BlockBasics but Refined Obsidian.

ToMe25 replied:
OK i got it.

thiakil commented:
I'm glad you worked out the light level stuff (before I had the chance to look into it).

Could you please give a bit more detailed explanation of you other changes?

ToMe25 commented:
What should i explain? why i did something or how i did it or what i did or something else?
What i did should be explained mostly in the PR Description.

thiakil commented:
The Why. Mainly the resistance changes, I'm not too concerned about Opaque/Full block, those should match the subtype's shape anyway.

ToMe25 commented:
OK.

I changed the resistance code to return `return blockResistance / 5.0F;` to match the Forge/Vanilla Code and return a Matching Value.

I changed the Refined Obsidian Resistance to 2400 as that is Obsidians Resistance x 2 as Obsidians Resistance should be 2000 * 3 / 5 = 1200 and some Changelog/Wiki(not sure which one) says the Refined Obsidian Resistance is twice the Obsidian Resistance.

I also changed the BlockBasic standard Resistance to 15 to not be twice the Iron Resistance but that realy doesn't does much of a difference, so if you want i can undo it, i just was thinking that might be why all BlockBasics are completely Explosion Resistant but it wasn't.

I changed the Refined Obsidian Hardness to the Obsidian Hardness just for logic and balance reasons.

And i changed some things to make the right Blocks Opaque but i won't explain that here as you sayed i don't need to.

ToMe25 commented:
I would btw suggest to change the Standard BlockBasic Resistance and Hardness to the Iron Block Values and the Osmium ones to Iron x 1.5 but what do you think about doing that?

pupnewfster commented:
I am fine with potentially increasing the hardness for various blocks, though I think in terms of resistance it makes more sense to update the documentation than to lower the value to match the documentation. This is because people's builds may rely on it having a specific blast resistance (who knows what people made for things)

pupnewfster commented on src/main/java/mekanism/common/block/BlockBasic.java:
```diff
@@ -497,6 +508,22 @@ public boolean isFullCube(IBlockState state) {
+        if(type != null) {
+            return type.isOpaqueCube ? 255 : 0;
+        }
+        return 0;
```
To be similar to how the other ones are, you want you could make this just be
```java
return type != null && type.isOpaqueCube ? 255 : 0;
```
Up to you though given it doesn't really matter, but once a decision is made and depending on if you are editing other things anyways you may want to.

ToMe25 replied:
btw i changed this and implemented the own Values for Osmium Blocks a few days ago and i will upload it until i got told what i should set the BlockBasicStandard Resistance to.

ToMe25 commented:
I think the Refined Obsidian Resistance doesn't realy changes anything as it should be completely blast resistant anyways but to what value do you mean to set the standard one? 20? 10? 100?
20 would be the old value but due to the Fix it would be less blast resistant the 100 would be the old value undoing the fix and 10 is the Iron Block one.
and than i think i will increase the Osmium Hardness.

pupnewfster commented:
Not sure, going to wait for @thiakil 's input. I also will relook up how the resistance gets used in terms of the formula and what other implications changing it may have, before I say anything final given I think I am remembering some of the things about it incorrectly (or at least not fully).

thiakil commented:
It would be good to know what documentation disagrees so we can work out if the doc was wrong to begin with, or the code has been changed, etc.

Throwing all these numbers around is getting a little confusing tbh. Could I trouble you to do a table/Google-Sheet of the current/proposed values plus a couple of relevant vanilla entries?

ToMe25 commented:
@thiakil i just figured it out later but the Refined Obsidian Resistance Change is more a Bugfix than a Change as the Value used for setResistance on Obsidian is 2000 and the Refined Obsidian Resistance was 2000 x 2 and not 2000 x 3 / 5 x 2 as it should be.
And i can make a table but not realy a google one, but what exactly should be in it?

thiakil commented:
> but what exactly should be in it?

The values that you're wanting to change and any relevant vanilla values (like iron, obsidian)

ToMe25 commented:
I think i got that table finished but its a zip as github dont likes csv files: 
[MekanismBlockBasicValues.csv.zip](https://github.com/mekanism/Mekanism/files/3228083/MekanismBlockBasicValues.csv.zip)

ToMe25 commented:
The Table in Markdown:

|                          | Refined-Obsidian | Osmium-Block | Every-Other-BlockBasic | Iron-Block | Obsidian | 
|--------------------------|------------------|--------------|------------------------|------------|----------| 
| setResistance-Value-New  | 4000             | 15           | 10                     | 10         | 2000     | 
| setResistance-Value-Old  | 6666⅔            | 20           | 20                     | 10         | 2000     | 
| Resistance-New           | 2400             | 9            | 6                      | 6          | 1200     | 
| Resistance-Old           | 4000             | 60           | 60                     | 6          | 1200     | 
| Resistance-Change-Bugfix | yes              | some-parts   | some-parts             | no         | no       | 
| Hardness-New             | 50               | 7.5          | 5                      | 5          | 50       | 
| Hardness-Old             | 5                | 5            | 5                      | 5          | 50       | 
| Hardness-Change-Bugfix   | no               | no           | no                     | no         | no       |

ToMe25 commented:
@thiakil I finished the Table now a Week ago, please tell me what i should use for the setResistance call in the Constructor.
I will make Osmium gets 5 more.

thiakil commented:
Apologies, I've been flat out with day-job work. Will try to get to it this weekend.

thiakil commented:
OK, Refinied obsidian stuff looks fine to be 2x Obsidian.

BlockBasic as a whole however doesn't really have a 'standard' value - each of the sub-blocks should be different. Bronze, Steel & Charcoal for example certainly don't have the same properties

ToMe25 commented:
OK but what should that Value for each one be? i would as a mentioned earlyer use 15 for Osmium but i think maybe 20 would be better and for Steel etc i dont realy know. also i think it would be good to have a standard value for example that could be used for new blocks.

ToMe25 commented:
I changed it to this now, do you think thats good or should i change some more values?

|                        |Refined-Obsidian|Osmium-Block|Bronze&Steel-Block|BlockBasicStandard|BlockBasicStandard-Bugfix-Only|Iron-Block|Obsidian|
|------------------------|------------|------------|------------|------------------|------------------------------|----------|--------|
|setResistance-Value-New |4000|20|15|10                |20                            |10        |2000    |
|setResistance-Value-Old |6666⅔|20|20|20                |20                            |10        |2000    |
|Resistance-New|2400|12|9|6                 |12                            |6         |1200    |
|Resistance-Old|4000|60|60|60                |60                            |6         |1200    |
|Resistance-Change-Bugfix|yes|some-parts|some-parts|some-parts        |yes                           |no        |no      |
|Hardness-New|50|7.5|5|5                 |5                             |5         |50      |
|Hardness-Old|5|5|5|5                 |5                             |5         |50      |
|Hardness-Change-Bugfix|no|no|no          |no                |no                            |no        |no      |

thiakil commented:
> use 15 for Osmium but i think maybe 20 would be better and for Steel

Sure, sounds fine. Steel you could look at what other mods do to ensure ours is similar

> Also i think it would be good to have a standard value for example that could be used for new blocks.

No new block are likely to be added to BlockBasic because we'll just have to flatten in 1.13/14, so no 'standard' should be needed

> Also i would give all the Thermal Evaporation Stuff the same Resistance

Sure, they're all made of Bronze IIRC

Security station should use whatever material it's made of, cant recall right now.

> And as Gold and Iron have the same Resistance i think it would be good to use the same value for Tin and Copper

Sure

ToMe25 commented:
I have looked for the Resistance of another Mods Steel Block and its 15 too so Steel Blocks, Bronze Blocks, the Thermal Evaporation Stuff and The Security Desk are now all have a setResistance of 15 Osmium Block has 20 and Standard is 10, should i upload it like that or change some values?

The Hardness of all Blocks but Osmium and Refined Obsidian is still 5.

thiakil commented:
upload away!

ToMe25 commented:
OK Uploaded that Changes.

thiakil commented on src/main/java/mekanism/common/block/BlockBasic.java:
```diff
@@ -270,9 +270,29 @@ public float getExplosionResistance(World world, BlockPos pos, Entity exploder,
+            return 2400F;
         }
-        return blockResistance;
+        else if (type == BasicBlockType.OSMIUM_BLOCK) {
```
This would probably look cleaner as a `switch`

ToMe25 replied:
Yeah it would, just didn't made one as i first only added a second if.
I will change that

ToMe25 commented:
btw what does this If?
There isn't a gui for Refined Obsidian right?
```java
    if (type == BasicBlockType.REFINED_OBSIDIAN) {
            if (entityplayer.isSneaking()) {
                entityplayer.openGui(Mekanism.instance, 1, world, pos.getX(), pos.getY(), pos.getZ());
                return true;
            }
        }
```

thiakil commented:
it is marked with `// Used to be credits UI` in `mekanism.client.ClientProxy#getClientGui`

ToMe25 commented:
but if i shift rightclick Refined Obsidian nothing happens

thiakil commented:
Thanks for all your work on this!

ToMe25 commented:
no problem :)
